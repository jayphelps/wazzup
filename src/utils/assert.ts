export const assert = (value: boolean, msg = 'Assertion failed') => {
  if (value === false) {
    throw new Error(msg);
  }
};
