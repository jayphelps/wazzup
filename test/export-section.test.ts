import { ast, wasm } from '../';
import { assertModule } from './utils/assertModule';

describe('Export Section', () => {
  let module;

  beforeEach(() => {
    module = new ast.Module();
  });

  test('adds exports', () => {
    const type1 = module.addFunctionType();
    const func1 = module.addFunction('foo', type1, [], []);

    module.addExport('foo', func1);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (func (;0;) (type 0))
      |   (export "foo" (func 0)))
    `);

    const type2 = module.addFunctionType();
    const func2 = module.addFunction('bar', type2, [], []);

    module.addExport('bar', func2);

    assertModule(module, `
      | (module
      |   (type (;0;) (func))
      |   (type (;1;) (func))
      |   (func (;0;) (type 0))
      |   (func (;1;) (type 1))
      |   (export "foo" (func 0))
      |   (export "bar" (func 1)))
    `);
  });
});
