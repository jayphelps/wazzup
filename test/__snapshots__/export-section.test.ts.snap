// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Export Section adds exports 1`] = `
Module {
  "exports": Array [
    Export {
      "externalKind": 0,
      "externalName": "foo",
      "value": Function {
        "body": Array [],
        "externalKind": 0,
        "functionType": FunctionType {
          "paramTypes": Array [],
          "returnTypes": Array [],
          "type": 96,
        },
        "localTypes": Array [],
        "name": "foo",
        "type": 112,
      },
    },
  ],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Export Section adds exports 2`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  4,
  1,
  96,
  0,
  0,
  3,
  2,
  1,
  0,
  7,
  7,
  1,
  3,
  102,
  111,
  111,
  0,
  0,
  10,
  4,
  1,
  2,
  0,
  11,
]
`;

exports[`Export Section adds exports 3`] = `
Module {
  "exports": Array [
    Export {
      "externalKind": 0,
      "externalName": "foo",
      "value": Function {
        "body": Array [],
        "externalKind": 0,
        "functionType": FunctionType {
          "paramTypes": Array [],
          "returnTypes": Array [],
          "type": 96,
        },
        "localTypes": Array [],
        "name": "foo",
        "type": 112,
      },
    },
    Export {
      "externalKind": 0,
      "externalName": "bar",
      "value": Function {
        "body": Array [],
        "externalKind": 0,
        "functionType": FunctionType {
          "paramTypes": Array [],
          "returnTypes": Array [],
          "type": 96,
        },
        "localTypes": Array [],
        "name": "bar",
        "type": 112,
      },
    },
  ],
  "functions": Array [
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "foo",
      "type": 112,
    },
    Function {
      "body": Array [],
      "externalKind": 0,
      "functionType": FunctionType {
        "paramTypes": Array [],
        "returnTypes": Array [],
        "type": 96,
      },
      "localTypes": Array [],
      "name": "bar",
      "type": 112,
    },
  ],
  "globals": Array [],
  "imports": Array [],
  "memories": Array [],
  "tables": Array [],
  "types": Array [
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
    FunctionType {
      "paramTypes": Array [],
      "returnTypes": Array [],
      "type": 96,
    },
  ],
  "version": 1,
}
`;

exports[`Export Section adds exports 4`] = `
Uint8Array [
  0,
  97,
  115,
  109,
  1,
  0,
  0,
  0,
  1,
  7,
  2,
  96,
  0,
  0,
  96,
  0,
  0,
  3,
  3,
  2,
  0,
  1,
  7,
  13,
  2,
  3,
  102,
  111,
  111,
  0,
  0,
  3,
  98,
  97,
  114,
  0,
  1,
  10,
  7,
  2,
  2,
  0,
  11,
  2,
  0,
  11,
]
`;
